modules/apps/27-interchain-accounts/keeper/relay_test.go

```diff
@@ -91,3 +100,128 @@ func (suite *KeeperTestSuite) TestTrySendTx() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestOnRecvPacket() {
+	var (
+		path       *ibctesting.Path
+		msg        sdk.Msg
+		txBytes    []byte
+		packetData []byte
+		sourcePort string
+	)
+
+	testCases := []struct {
+		msg      string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"Interchain account successfully executes banktypes.MsgSend", func() {
+				// build MsgSend
+				amount, _ := sdk.ParseCoinsNormalized("100stake")
+				interchainAccountAddr, _ := suite.chainB.GetSimApp().ICAKeeper.GetInterchainAccountAddress(suite.chainB.GetContext(), path.EndpointA.ChannelConfig.PortID)
+				msg = &banktypes.MsgSend{FromAddress: interchainAccountAddr, ToAddress: suite.chainB.SenderAccount.GetAddress().String(), Amount: amount}
+				// build packet data
+				txBytes, err := suite.chainA.GetSimApp().ICAKeeper.SerializeCosmosTx(suite.chainA.Codec, msg)
+				suite.Require().NoError(err)
+
+				data := types.IBCAccountPacketData{Type: types.EXECUTE_TX,
+					Data: txBytes}
+				packetData = data.GetBytes()
+			}, true,
+		},
+		{
+			"Cannot deserialize txBytes", func() {
+				txBytes = []byte("invalid tx bytes")
+				data := types.IBCAccountPacketData{Type: types.EXECUTE_TX,
+					Data: txBytes}
+				packetData = data.GetBytes()
+			}, false,
+		},
+		{
+			"Cannot deserialize txBytes: invalid IBCTxRaw", func() {
+				txBody := []byte("invalid tx body")
+				txRaw := &types.IBCTxRaw{
+					BodyBytes: txBody,
+				}
+
+				txBytes = suite.chainB.Codec.MustMarshal(txRaw)
+				data := types.IBCAccountPacketData{Type: types.EXECUTE_TX,
+					Data: txBytes}
+				packetData = data.GetBytes()
+			}, false,
+		},
+		{
+			"Invalid packet type", func() {
+				txBytes = []byte{}
+				// Type here is an ENUM
+				// Valid type is types.EXECUTE_TX
+				data := types.IBCAccountPacketData{Type: 100,
+					Data: txBytes}
+				packetData = data.GetBytes()
+			}, false,
+		},
+		{
+			"Cannot unmarshal interchain account packet data into types.IBCAccountPacketData", func() {
+				packetData = []byte{}
+			}, false,
+		},
+		{
+			"Unauthorised: Interchain account not found for given source portID", func() {
+				sourcePort = "invalid-port-id"
+			}, false,
+		},
+		{
+			"Unauthorised: Signer of message is not the interchain account associated with sourcePortID", func() {
+				// build MsgSend
+				amount, _ := sdk.ParseCoinsNormalized("100stake")
+				// Incorrect FromAddress
+				msg = &banktypes.MsgSend{FromAddress: suite.chainB.SenderAccount.GetAddress().String(), ToAddress: suite.chainB.SenderAccount.GetAddress().String(), Amount: amount}
+				// build packet data
+				txBytes, err := suite.chainA.GetSimApp().ICAKeeper.SerializeCosmosTx(suite.chainA.Codec, msg)
+				suite.Require().NoError(err)
+				data := types.IBCAccountPacketData{Type: types.EXECUTE_TX,
+					Data: txBytes}
+				packetData = data.GetBytes()
+			}, false,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
```

Comment:
```suggestion
		suite.Run(tc.msg, func() {
```
This is what I'll usually do, but happy to adjust if it is useful to prefix with `Case `